lib/resque.rb

```diff
@@ -181,11 +180,35 @@ def push(queue, item)
   #           indeterminate blocking.
   #
   # Returns an array of [queue_name, decoded_payload], or falsey.
-  def pop(queues, timeout=1)
+  def pop(queues, timeout = 1, retries = 3)
     queue_names = Array(queues).map { |queue| "queue:#{queue}" }
     timeout = [1, timeout].max # require nonzero, no infinite blocking
-    queue, payload = redis.blpop(*queue_names, timeout)
-    queue && [queue.sub("queue:", ""), decode(payload)]
+    begin
+      queue, payload = redis.blpop(*queue_names, timeout)
+      queue && [queue.sub("queue:", ""), decode(payload)]
+    rescue Redis::TimeoutError => e
+      # This exception happens when Redis goes
+      # away or we failover through a proxy layer.
+      if reconnect(retries)
+        # Wait for a random time between 0 and 1 second to prevent thundering reconnect herd
+        sleep rand
```

Comment:
Is this second sleep (the first being within `#reconnect`) necessary? If it's successfully reconnected, should we wait any longer to resume operation?

Reply:
Good catch, this was intended to be above the `reconnect` but I had been shuffling around the code a few times. Moving it.